types/wlr_drm_lease_v1.c

```diff
@@ -0,0 +1,680 @@
+
+		if (connector == tmp) {
+			wlr_log(WLR_ERROR, "Connector has already been requested");
+			return;
+		}
+	}
+
+	size_t n_connectors = request->n_connectors + 1;
+
+	struct wlr_drm_lease_connector_v1 **tmp_connectors =
+			realloc(request->connectors,
+			n_connectors * sizeof(struct wlr_drm_lease_connector_v1 *));
+	if (!tmp_connectors) {
+		wlr_log(WLR_ERROR, "Failed to grow connectors request array");
+		return;
+	}
+
+	request->connectors = tmp_connectors;
+	request->connectors[request->n_connectors] = connector;
+	request->n_connectors = n_connectors;
+}
+
+static void drm_lease_request_v1_handle_submit(
+		struct wl_client *client, struct wl_resource *resource, uint32_t id) {
+	struct wlr_drm_lease_request_v1 *request =
+			drm_lease_request_v1_from_resource(resource);
+
+	struct wlr_drm_lease_v1 *lease = calloc(1, sizeof(struct wlr_drm_lease_v1));
+	if (!lease) {
+		wlr_log(WLR_ERROR, "Failed to allocate wlr_drm_lease_v1");
+		wl_resource_post_no_memory(resource);
+		return;
+	}
+
+	uint32_t version = wl_resource_get_version(resource);
+	struct wl_resource *lease_resource = wl_resource_create(client,
+			&wp_drm_lease_v1_interface, version, id);
+	if (!lease_resource) {
+		wlr_log(WLR_ERROR, "Failed to allocate wl_resource");
+		wl_resource_post_no_memory(resource);
+		free(lease);
+		return;
+	}
+
+	lease->device = request->device;
+	lease->resource = lease_resource;
+	wl_list_insert(&lease->device->leases, &lease->link);
+
+	request->lease = lease;
+
+	wl_resource_set_implementation(lease_resource, &lease_impl,
+			lease, drm_lease_v1_handle_resource_destroy);
+
+	if (request->invalid) {
+		/* Pre-emptively reject invalid lease requests */
+		wlr_log(WLR_ERROR, "Invalid request");
+		wp_drm_lease_v1_send_finished(lease->resource);
+		drm_lease_v1_destroy(lease);
+	} else {
+		/* TODO: reject the request if the user does not grant it */
+		wlr_signal_emit_safe(&request->device->manager->events.request,
+				request);
+	}
+
+	/* Request is done */
+	wl_resource_destroy(resource);
+}
+
+static struct wp_drm_lease_request_v1_interface lease_request_impl = {
+	.request_connector = drm_lease_request_v1_handle_request_connector,
+	.submit = drm_lease_request_v1_handle_submit,
+};
+
+static void drm_lease_device_v1_handle_resource_destroy(
+		struct wl_resource *resource) {
+	wl_list_remove(wl_resource_get_link(resource));
+}
+
+static void drm_lease_device_v1_handle_stop(
+		struct wl_client *client, struct wl_resource *resource) {
+	wp_drm_lease_device_v1_send_finished(resource);
+	wl_resource_destroy(resource);
+}
+
+static void drm_lease_device_v1_handle_create_lease_request(
+		struct wl_client *client, struct wl_resource *resource, uint32_t id) {
+	struct wlr_drm_lease_device_v1 *device =
+		drm_lease_device_v1_from_resource(resource);
+	if (!device) {
+		wlr_log(WLR_ERROR, "Failed to create lease request, "
+				"wlr_drm_lease_device_v1 has been destroyed");
+		wl_resource_post_no_memory(resource);
```

Comment:
It's incorrect to send a protocol error here: `drm_lease_device_v1_from_resource` returns NULL when the resource is inert. This happens when the compositor has destroyed the device but the client isn't aware of this yet. So it's not a client error, the client cannot do anything to prevent this race. We shouldn't punish the client and kill their connection here.

Instead, we need to create an inert lease request.

* * *

To simplify the logic, it may be easier to unconditionally create an inert lease request, and only turn it into a real lease request if we have a device.

```c
struct wl_resource *request_resource = wl_resource_create(client,
	&wp_drm_lease_request_v1_interface, version, id);
wl_resource_set_implementation(request_resource, &lease_request_impl,
	NULL, drm_lease_request_v1_handle_resource_destroy);
if (!device) {
	return; // Leave the lease request inert
}

struct wlr_drm_lease_request_v1 *req = calloc(1, sizeof(struct wlr_drm_lease_request_v1));
// Initialize req here

// Turn the inert lease request into a live lease request
wl_resource_set_user_data(request_resource, req);
```

A similar trick could be used to simplify `drm_lease_request_v1_handle_submit` too.

Reply:
Should we still create the `request_resource` if nothing is sent to the client?

Reply:
We should, otherwise the connection will be killed when the client tries to use the non-existing object.